Add i18n doc (#1904)
spryker/spryker-docs: docs/scos/dev/front-end-development/202212.0/oryx/oryx-localization.md

---
title: "Oryx: Localization"
description: Localizations are decoupled from component implementations
template: concept-topic-template

> last_updated: July 23, 2023

Reviewer: ```suggestion
last_updated: Aug 23, 2023

```

---

Localization is the part of internationalization (`i18n`) that's concerned with creating variations of the same application in different locales.

While most localized content is managed in other systems, like product or content management systems, components often use small text fragments in the UI. The examples are static site labels or aria labels.

To make sure all components can be customized and localized, Oryx doesn't provide _any_ hardcoded text. Instead, components use _tokens_ that function as a reference to a text. When there's no localized text available for such a reference, Oryx renders the token in a human-readable format.

## Translation keys

> Translation keys, also knows as _i18n tokens_, are used to resolve localized text. The tokens are created with the following conventions in mind:

Reviewer: ```suggestion
Translation keys, also known as _i18n tokens_, are used to resolve localized text. The tokens are created with the following conventions in mind:

```

- Tokens are written in English.
- Tokens are written in kebab-case format—for example, `my-token`.
- Tokens are organized by domains—for example, `cart.add-to-cart`.
- Tokens support context variables, which are wrapped in angle brackets in camelCase format. For example, `cart.totals.<count>-items`.

## Resolving translations from translation resources

> The localization of labels is driven by the current language and the translation key. Translations are typically provided as lazy-loaded resources next to the component implementation. Although, they can aslo be added as part of the static resources that are loaded in Oryx. If the resources are loaded externally, they can be provided by static JSON files or a third-party service through an API.

Reviewer: ```suggestion
The localization of labels is driven by the current language and the translation key. Translations are typically provided as lazy-loaded resources next to the component implementation. However, they can also be added as part of the static resources that are loaded in Oryx. If the resources are loaded externally, they can be provided by static JSON files or a third-party service through an API.

```

> The localization of labels is driven by the current language and the translation key. Translations are typically provided as lazy-loaded resources next to the component implementation. Although, they can aslo be added as part of the static resources that are loaded in Oryx. If the resources are loaded externally, they can be provided by static JSON files or a third-party service through an API.

Reviewer 2: However, they can also be added as part of the static resources loaded in Oryx. If the resources are loaded externally, they can be provided by static JSON files or a third-party service through an API.

> The localization of labels is driven by the current language and the translation key. Translations are typically provided as lazy-loaded resources next to the component implementation. Although, they can aslo be added as part of the static resources that are loaded in Oryx. If the resources are loaded externally, they can be provided by static JSON files or a third-party service through an API.

Reviewer 2: ```suggestion
The localization of labels is driven by the current language and the translation key. Translations are typically provided as lazy-loaded resources next to the component implementation. Although, they can aslo be added as part of the static resources that are loaded in Oryx. If the resources are loaded externally, they can be provided by static JSON files or a third-party service through an API.

```
However, they can also be added as part of the static resources loaded in Oryx. If the resources are loaded externally, they can be provided by static JSON files or a third-party service through an API.

Oryx uses the active application language to look up the available labels. When the language is `en`, the locales for English are resolved.

If a language resource can be resolved, the translation key is evaluated against the available translations in the resource using all "parts" of the key. For example, if a translation key contains multiple parts, like `cart.increase`, you can provide a global translation for just the `increase` part or for `cart.increase`.

> ## Auto-conversion of translation keys

Reviewer: ```suggestion
## Autoconversion of translation keys

```

| `cart.totals.<count>-items` | 5 items         |

This mechanism lets Oryx avoid distributing localizations as a standard package or as part of the boilerplate. Labels are quite opinionated and may change frequently over time, which may cause breaking changes. For 90% of the cases, the keys are fairly short and provide an OK experience.

{% info_block infoBox "" %}

The [Oryx labs package](https://www.npmjs.com/package/@spryker-oryx/labs) provides some localizations mainly for demonstration reasons. For more information on the labs package, see [Feature sets](/docs/scos/dev/front-end-development/{{page.version}}/oryx/feature-sets.html).

{% endinfo_block %}

## Installing the i18n package

To use the i18n package, add `I18nFeature` to the app and configure translation resources using the loader function:

```ts
import { appBuilder } from "@spryker-oryx/application";
import { I18nFeature } from "@spryker-oryx/i18n";

export const app = appBuilder()
  .withFeature(
    new I18nFeature({
      // Here as example we are loading translations from TS modules
      load: (localeId) => import(`../i18n/${localeId}.ts`), // <-- Required part
    })
  )
  .create();
```

> This gives you full flexibility of how and where to load your translation texts from. The `load` function returns a promise of a `{ default: I18nData }` type.

Reviewer: ```suggestion
This gives you full flexibility in how and where to load your translation texts. The `load` function returns a promise of a `{ default: I18nData }` type.

```

  // pluralization example
  "order.<count>-items":
    "Products ({count, plural, one {{count} item} other {{count} items}})",
};
{% endraw %}```

You can also load the text from a third-party translation engine.

The [globalize](https://www.npmjs.com/package/globalize) translation engine is based on the standardized [ICU message expressions](https://unicode-org.github.io/icu/userguide/format_parse/messages/). The ICU message format defines a standard syntax to handle common translation cases:

- Pluralization: for example, `You have {count} {count, plural, one {item} other {items}} in cart`.
- Selects: for example, `{gender, select, male {He} female {She} other {They}} invited you to party!`.

Other ICU capabilities, such as [number formatting](https://unicode-org.github.io/icu/userguide/format_parse/numbers/) or [date and time formatting](https://unicode-org.github.io/icu/userguide/format_parse/datetime/) can be loaded in addition. Most likely, it's not needed in your project, because Oryx uses the standard [Intl](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl) API, which is widely supported.

## Using the i18n directive in components

> Localized content uses [the fine-grained reactivity system](/docs/scos/dev/front-end-development/{{page.version}}/oryx/reactivity/reactivity.html). The localizations are updated as soon the active locale is changed; for example, when the user interacts with the localization selector, or when the `LocaleService.set()` API is used. All localized content in Oryx components is updated instantly without a page reload.

Reviewer: ```suggestion
Localized content uses [the fine-grained reactivity system](/docs/scos/dev/front-end-development/{{page.version}}/oryx/reactivity/reactivity.html). The localizations are updated as soon the active locale is changed; for example, when the user interacts with the localization selector or when the `LocaleService.set()` API is used. All localized content in Oryx components is updated instantly without a page reload.

```
